### PR TITLE
[T0071-GA < DFS] add dfs and tests

### PR DESCRIPTION
**Pull request by brunos252**
### Description

This is a copy-paste of #142, replacing `*bfs` with `*` and making DepthFirstSearch instead of BreadthFirstSearch.

### Pull request type

- [x] Feature

### Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

######################################

### Reviewer checklist (the reviewer checks this part)
- [ ] Core feature implementation
- [ ] Tests
- [ ] Code documentation
- [ ] Documentation on [memgraph/docs](https://github.com/memgraph/docs)

######################################

### CODE REVIEW

File: tests/memgraph/test_query_builder.py [hunk 1]
```diff
@@ -1340,3 +1340,57 @@ class Follows_test(Relationship, type="FOLLOWS"):
...
+
+    expected_query = " MATCH (:City {name: 'Paris'})-[:Road * (e, v | e.length <= 200 AND v.name != 'Metz')]->(:City {name: 'Berlin'}) RETURN * "
+
+    with patch.object(Memgraph, "execute_and_fetch", return_value=None) as mock:
+        query_builder.execute()
+
+    mock.assert_called_with(expected_query)
+
+
+def test_dfs_bounds():
```
Comment (katarinasupe) [inline tests/memgraph/test_query_builder.py:1384]: Add tests for `lower_bound` and all combinations.

**Review by katarinasupe — CHANGES_REQUESTED** [review]
Similar comments apply to this PR as in BFS.

**Review by katarinasupe — APPROVED** [review]

File: gqlalchemy/memgraph.py [hunk 2]
```diff
@@ -586,3 +586,50 @@ def to_cypher_lambda(expression: str) -> str:
...
+    def __init__(
+        self,
+        lower_bound: int = None,
+        upper_bound: int = None,
+        condition: str = None,
+    ) -> None:
+        super().__init__()
+        self.lower_bound = lower_bound
+        self.upper_bound = upper_bound
+        self.condition = condition
```
Comment (Josipmrden) [inline gqlalchemy/memgraph.py:611]: Explain the arguments in pydocs

File: gqlalchemy/memgraph.py [hunk 3]
```diff
@@ -586,3 +586,50 @@ def to_cypher_lambda(expression: str) -> str:
...
+        algo_str = " *"
+
+        bounds = self.to_cypher_bounds()
+        if bounds != "":
+            algo_str += f" {bounds}"
+
+        filter_lambda = super().to_cypher_lambda(self.condition)
+        if filter_lambda != "":
+            algo_str += f" {filter_lambda}"
+
```
Comment (Josipmrden) [inline gqlalchemy/memgraph.py:624]: algo_str = f"{algo_str} {filter_lambda}"

Comment (Josipmrden) [reply gqlalchemy/memgraph.py:624]: Same for above in bounds

File: gqlalchemy/memgraph.py [hunk 4]
```diff
@@ -586,3 +586,50 @@ def to_cypher_lambda(expression: str) -> str:
...
+        return algo_str
+
+    def to_cypher_bounds(self) -> str:
+        """If bounds are specified, returns them in grammar-defined form."""
+        if self.lower_bound is None and self.upper_bound is None:
+            return ""
+
+        lower = str(self.lower_bound) if self.lower_bound is not None else ""
+        upper = str(self.upper_bound) if self.upper_bound is not None else ""
+
```
Comment (Josipmrden) [inline gqlalchemy/memgraph.py:634]: Good one, just keep doing ternary operators when double choiced! :+1:

**Review by Josipmrden — COMMENTED** [review]
Minor comments

File: gqlalchemy/memgraph.py [hunk 5]
```diff
@@ -580,3 +582,57 @@ def to_cypher_lambda(expression: str) -> str:
...
+            algo_str = f"{algo_str} {filter_lambda}"
+
+        return algo_str
+
+    def to_cypher_bounds(self) -> str:
+        """If bounds are specified, returns them in grammar-defined form."""
+        if self.lower_bound is None and self.upper_bound is None:
+            return ""
+
+        lower = str(self.lower_bound) if self.lower_bound is not None else ""
```
Comment (BorisTasevski) [inline gqlalchemy/memgraph.py:635]: Could we move this logic to ctor?

File: gqlalchemy/memgraph.py [hunk 6]
```diff
@@ -580,3 +582,57 @@ def to_cypher_lambda(expression: str) -> str:
...
+        upper = str(self.upper_bound) if self.upper_bound is not None else ""
```
Comment (BorisTasevski) [inline gqlalchemy/memgraph.py:636]: Could we move this logic to ctor?

**Review by BorisTasevski — CHANGES_REQUESTED** [review]

**Review by BorisTasevski — APPROVED** [review]

**Pull request merged**